third_party/move/tools/move-package-manifest/src/util.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use annotate_snippets::{Level, Renderer, Snippet};
+use std::fmt::Write;
+
+/// Renders a formatted error message for a toml deserialization failure, optionally with
+/// source span annotations.
+///
+/// If the error contains span information, this function highlights the relevant portion
+/// of the original text using a diagnostic-style output. Otherwise, it falls back to a
+/// plain error message.
+///
+/// # Example Output
+///
+/// ```text
+/// error: failed to parse manifest
+///   |
+/// 1 | [package]
+/// 2 | name = "some_package_name"
+/// 3 | version = "0.1.2"
+/// 4 | upgrade_policy = "invalid-policy"
+///   |                  ^^^^^^^^^^^^^^^^ unknown variant `invalid-policy`, expected `compatible` or `immutable`
+///   |
+/// ```
+pub fn render_error(
+    output: &mut impl Write,
+    manifest_text: &str,
+    err: &toml::de::Error,
+) -> anyhow::Result<()> {
+    match err.span() {
+        Some(span) => {
+            let snippet = Snippet::source(manifest_text)
+                .annotation(Level::Error.span(span).label(err.message()));
+
+            writeln!(
```

Comment:
Nit: Can the manifest files ever be very large? In this case, it seems like printing out the entire manifest (as opposed to just the part with error) may be cumbersome for user.

Reply:
The manifests are likely to remain reasonably small in the foreseeable future as we have a lot fewer options than Rust and Move projects likely have fewer dependencies, so I think it's fine to render the whole stuff for now.

Should future needs arise we can improve this easily.